KEP-4033: Discover cgroup driver from CRI

marquiz opened the pull request:
<!-- 
	Please use the following format when naming your PR
	< Issue Number >:< Issue Description >
	e.g. KEP-000: adding beta graduation criteria
	
	Avoid using phrases like `fixes #NNNN` in the description
	unless the pull request is to change the KEP status to 
	implemented or KEP has been deprecated.
-->

<!-- short description of work done in PR e.g. updating milestone, adding new KEP, adding test requirements… -->  
- One-line PR description: KEP for discovering kubelet cgroup driver from CRI

<!-- link to the k/enhancements issue -->
- Issue link: #4033 

<!-- other comments or additional information -->
- Other comments:

marquiz commented:
/cc haircommander

adisky commented on keps/sig-node/4033-group-driver-detection-over-cri/README.md:
```diff
@@ -0,0 +1,880 @@
+The existing RuntimeStatus message (of the existing Status API endpoint) is
+used as this is being frequently queried by the kubelet, and is a place where
+the runtime tells the Kubelet about its state. The runtime will decide which
+CgroupDriver to choose based on existing methods: its own configuration.
```
Does kubelet have any kinds of cgroup interaction before it gets response from `RuntimeStatus`?  possibly creation of cgroup for its own usage or QOS hierarchy?

adisky commented on keps/sig-node/4033-group-driver-detection-over-cri/README.md:
```diff
@@ -0,0 +1,880 @@
+This enhancement adds the ability for the container runtime to tell kubelet
+which cgroup driver to use. This removes the need for specifying cgroup driver
+in the kubelet configuration and eliminates the possibility of misaligned
+cgroup driver configuration between the kubelet and the runtime.
```
Are we planning to support cgroupfs driver for cgroupsv2 as well? If not then by the time this KEP accepted and reached to GA, we might start deprecating cgroupsv1

I am not sure about timelines but this could be a possibility

marquiz replied:
Are you talking about kubelet or the runtimes? I think kubelet at least supports cgroupfs for cgroupsv2, too(?)

BenTheElder replied:
Worth considering that both the runtimes and kubelet have to be updated to land this in which time we might just only be doing systemd, @mrunalp @dchen1107 thoughts?

mrunalp replied:
I think that is a fair point we are moving towards a systemd cgroup driver default. Do we have a need to make something like this more general? Are there other features that kubelet would want to discover from the runtime?

haircommander replied:
> maybe replaced with GC pinning on the CRI implementation side?

yeah the intention is this

in sig node we discussed this and also came up with QoS class, runtime class and CRI features (like CRI saying it supports user namespaces). However, for now, we opted to leave this KEP only focusing on cgroup driver

marquiz replied:
Let's keep the possible future extensions in mind, as @mrunalp mentioned, keep it generalizable

BenTheElder replied:
Seems like we can mark this thread resolved? :-)

aojea commented:
/assign @BenTheElder

ffromani commented:
/cc

marquiz commented:
API proposal updated. It now has `LinuxRuntimeConfiguration` sub-type under which the cgroup driver info resides. It's possible to add corresponding windows sub-type (and common(?)) in the future

bart0sh commented:
/cc

neolit123 commented:
cc @SataQiu @pacoxu @chendave

neolit123 commented:
thanks for taking a stab at this. for the record, this has been the most common, yet rather unclear to users, error for low level setups such as kubeadm.

sftim commented on keps/sig-node/4033-group-driver-detection-over-cri/README.md:
```diff
@@ -0,0 +1,890 @@
+
+## Summary
+
+This enhancement adds the ability for the container runtime to tell kubelet
+which cgroup driver to use. This removes the need for specifying cgroup driver
+in the kubelet configuration and eliminates the possibility of misaligned
+cgroup driver configuration between the kubelet and the runtime.
```
> tell kubelet which cgroup driver to use

Is this an instruction, or a hint?

marquiz replied:
an instruction -> reworded to _instructed_

sftim commented on keps/sig-node/4033-group-driver-detection-over-cri/README.md:
```diff
@@ -0,0 +1,890 @@
+
+### Goals
+
+- make kubelet automatically use the same cgroup driver as the container
```
``suggestion
- allow kubelet to automatically use the same cgroup driver as the container
```
?

(I think we'd support an override mechanism, but maybe I'm wrong)

marquiz replied:
No we don't. There should be no reason to override as the cgroup driver setting of and the kubelet and runtime really must be aligned (the same). Anything else is misconfiguration and effectively a "broken" (partially or fully) node

neolit123 replied:
agreed. it may make some sense to let a a CR expose cgroup driver config to the user. but the kubelet cgroup driver CLI flag and v1beta1 config option must be removed eventually. it should be automatic

to me, it makes sense to outline this deprecation work in the kep and then log an issue in k/k for sig node to track. but(!!) these cli/config options are so widely used that inevitably, and despite release notes, the removal will break some users.

haircommander replied:
> it may make some sense to let a CR expose cgroup driver config to the user

they could potentially do this by having a CR that configures the CRI implementation, but I think the larger intention of this enhancement is to take the configuration of cgroup driver away from kubelet+kubernetes in general

marquiz replied:
Proposal updated: now the "override" mechanism would be to disable the feature gate. @sftim this resolved?

sftim replied:
I still recommend the wording change, unless there's a reason it makes the KEP less useful

marquiz replied:
Hmm, might be hair-splitting but I feel that "allow" changes the aim of this proposal. The goal is for kubelet to obey what runtime tells it to do (when the feature is enabled), not to opt out for something else. Maybe it's my non-native language skills... WDYT @haircommander @sftim ?

sftim commented on keps/sig-node/4033-group-driver-detection-over-cri/README.md:
```diff
@@ -0,0 +1,890 @@
+### Risks and Mitigations
+
+Field adoption could be considered a risk, though the CRI implementations work
+closely with SIG-Node and the feature will move along with CRI implementation
```
(nit)
```suggestion
closely with SIG Node and the feature will move along with CRI implementation
```

marquiz replied:
Thx, fixed

sftim commented on keps/sig-node/4033-group-driver-detection-over-cri/README.md:
```diff
@@ -0,0 +1,890 @@
+information about the cgroup driver, then kubelet will fall back to using its
+own configuration (`cgroupDriver` from kubeletConfig or the `--cgroup-driver`
+flag). Further, the kubeletConfig field and `--cgroup-driver` flag will be
+marked as deprecated, to be dropped when the runtimes the Kubelet is supported
+to run with all support the flag.
```
AIUI we do not publish a list of supported vs. unsupported container runtimes.

(We do tell people what we test against, but I don't think we - Kubernetes - make a public claim about compatibility with some runtimes and not others).

marquiz replied:
True. Though in practice we only have a few implementations. We could drop the deprecation part of the proposal, as the flag is still valid and necessary with runtime versions that doesn't support this. Any thoughts @haircommander?

haircommander replied:
We could also be explicit in what we mean by all. Members of the cri-o and containerd community have signed-off on the proposal. I am not aware of cri-dockerd's awareness. We could change the wording to
```
to be dropped when support for the field is adopted by CRI-O and containerd.
```

haircommander replied:
tbh I think we should deprecate so users have warning that they should remove it from their scripts and the like. I don't think we need to be explicit about when we will drop it.

```
The --cgroup-driver option has been deprecated and will be dropped in a future release. Please upgrade to a CRI implementation that supports cgroup-driver detection.
```
or something

marquiz replied:
OK, added back with the following wording:

```
Further, the kubeletConfig field and `--cgroup-driver` flag will be
marked as deprecated, to be dropped when support for the feature is adopted by
CRI-O and containerd. Usage of the deprecated setting will produce a log
message, e.g.:

"cgroupDriver option has been deprecated and will be dropped in a future release. Please upgrade to a CRI implementation that supports cgroup-driver detection."
```

sftim commented on keps/sig-node/4033-group-driver-detection-over-cri/README.md:
```diff
@@ -0,0 +1,890 @@
+logs or events for this purpose.
+-->
+
+Kubelet and CRI version.
```
If I know these values, and have no information about the container runtime, can I work out whether or not that the kubelet is automatically learning which cgroup driver to use?

marquiz replied:
Woops, this should actually state "container runtime" instead of "CRI". PR updated accordingly.

The end user doesn't need to know about this. The node admin ahould know what version of container runtime is in use. @haircommander

sftim replied:
For Kubernetes, the node admin is usually _our_ end user. We use the term “cluster operator” in docs (which does sometimes confuse, but that's what we have in the style guide).

marquiz replied:
Ach, yes of course, thanks for correcting me. But anyway, the answer would be "Kubelet and container runtime version"

marquiz replied:
This now states:

_Kubelet and container runtime version. The [`crictl`](https://kubernetes.io/docs/tasks/debug/debug-cluster/crictl/) tool can be used to determine if the container runtime supports the feature (`crictl info`)._

@sftim do you consider this as resolved?

marquiz replied:
Resolving this comment as a duplicate of https://github.com/kubernetes/enhancements/pull/4034#discussion_r1217944720

sftim commented on keps/sig-node/4033-group-driver-detection-over-cri/README.md:
```diff
@@ -0,0 +1,890 @@
+Recall that end users cannot usually observe component logs or access metrics.
+-->
+
+No metrics likely will expose this.
```
How about exposing a cgroup driver time series? It'd be low cardinality I think.

marquiz replied:
In practice, this setting not expected to change. Metrics for configuration settings deserves a separate proposal, if desired, I think. @haircommander

sftim replied:
If we're loading it from third party software, then surely it's not a configuration setting?

(imagine if you update to a new system image for nodes with a different container runtime, and now you have a mix of cgroup v2 and the future cgroup v3 - you might care)

haircommander replied:
yeah I guess I also find myself wondering how useful such a metric would be. cgroup version (1, 2, or future 3) is not covered in the scope of this KEP--it's automatically discovered by both kubelet and cri. I don't see any reason a cluster admin would want a metric to see the cgroup driver.

mrunalp replied:
How about displaying it in Node Status?

marquiz replied:
Would be possible, of course, but I'd like to keep Kubernetes API changes out of the scope of this KEP if possible.

sftim commented on keps/sig-node/4033-group-driver-detection-over-cri/README.md:
```diff
@@ -0,0 +1,890 @@
+inform the runtime which cgroup driver to use. This could be achieved e.g.
+without any changes to the CRI API by the CRI implementation guessing the
+cgroup driver based on the path of the CgroupParent of the pod, passed down in
+the RunPodSandboxRequest. However, SIG-Node has decided that the CRI
```
```suggestion
the RunPodSandboxRequest. However, SIG Node has decided that the CRI
```

marquiz replied:
Fixed

sftim commented on keps/sig-node/4033-group-driver-detection-over-cri/kep.yaml:
```diff
@@ -0,0 +1,39 @@
+
+# The following PRR answers are required at alpha release
+# List the feature gate name and the components for which it must be enabled
+feature-gates: []
```
As well as changing the CRI protocol, I'd expect to see a feature gate (even one that's on by default) to control whether the kubelet falls back to manual configuration or honors what the container runtime reports.

sftim replied:
> If available the cgroup driver information received from the container runtime  will take precedence over cgroupDriver setting from the kubelet config (or `--cgroup-driver` command line flag).

It doesn't sound like there's any other way to opt out. In that case we might want to mark the feature as beta from the off.

marquiz replied:
I really can't think of a scenario where manual override would be desired. If the settings are not aligned the node is misconfigured and not working properly. This is exactly the (commmon'ish) misconfiguration scenario what we're trying to prevent with the proposal. You're right with the straight to beta 🤔 Thoughts @haircommander?

haircommander replied:
personally, I don't think a feature gate is necessary. I think the kubelet should be robust enough to fallback to its own cgroup-driver for a couple of releases while broad adoption of newer cri's takes place. Eventually, it can drop support for its own cgroup-driver. we shouldn't drop the flag until we're confident the runtimes are present.

> It doesn't sound like there's any other way to opt out

I don't think we should provide a way to opt-out tbh. the runtime already needs to be configured manually for a cgroup driver, so the cluster admin already needs to know that cgroup-driver should be configured. The only difference is the cluster admin now has one place fewer to do that configuration.

aojea replied:
feature flags are not meant to be used as opt-in or opt-out flags, they are meant to graduate and eventually being removed. They are used to give the project the opportunity to roll back changes on multiple releases, if something unexpected happen on this feature and we introduce this change directly, we'll be releasing a core component with a bug that we can only solve with a new release.

As Tim said, when we are confident with the change sometimes we start with a feature gate enabled by default in beta, but this change is introducing a behavior change, it must have a feature gate and provide a graduation criteria (it should work for containerd and crio per example and no bugs reported during one release, per example) IMHO

mrunalp replied:
Another thing that a feature gate can help with is if we need to change implementation details around how we merge or override we are protected and can stay in alpha to do that.

marquiz replied:
OK, looks like people want this so I added feature gate (`KubeletCgroupDriverFromCRI`) in the proposal. It is proposed to be enabled by default, and thus KEP maturity would be beta directly. @sftim @aojea @mrunalp did I get it right?

bart0sh commented on keps/sig-node/4033-group-driver-detection-over-cri/README.md:
```diff
@@ -0,0 +1,890 @@
+
+The responsibility of managing the Linux cgroups is currently split between the
+kubelet and the container runtime. Kubelet takes care of the pod (sandbox)
+level cgroups whereas the runtime is responsible for per-container cgroups.
```
Probably silly question. Why don't we want to delegate sandbox cgroups management to a runtime? If this is doable we don't need to synchronise this between Kubelet and runtime anymore.

sftim replied:
Is it because we also want to control the namespaces that are associated with things in the sandbox cgroup?

marquiz replied:
I don't think it's a silly question. I think we'll get back to that at some other KEP :) That's just soo much bigger topic and "problem" to tackle. And would take time to adopt even if we had the code merged in the runtimes' mainline now.

marquiz replied:
> Is it because we also want to control the namespaces that are associated with things in the sandbox cgroup?

I think it's for historical reasons (docker) but others probably know better than me.

sftim commented on keps/sig-node/4033-group-driver-detection-over-cri/README.md:
```diff
@@ -0,0 +1,890 @@
+logs or events for this purpose.
+-->
+
+Kubelet and container runtime version.
```
If I know these version values, and I **don't** have a detailed understanding of the behavior of my preferred container runtime, can I work out whether or not the kubelet is automatically learning which cgroup driver to use?

If not, I'd like us to add that. It's much easier for us to document “look at this metrics query” or “look for this log line” vs. “go and read the docs and / or source code for some third party component”.

We should add this for the first release where the new behavior will be on by default, and optionally for any earlier releases.

This question is not the same as can I work out which driver is in use; it's about working out whether or not the kubelet is overriding my configuration. (It is also useful to know which driver is in use).

haircommander replied:
would having the output of `crictl info` display the cgroupDriver be sufficient? it currently reports the runtime status info and would likely be extended to include it, which would provide a runtime agnostic way of polling

marquiz replied:
I added a mention of `crictl info` for determining the runtime-side support. WDYT @sftim

marquiz replied:
This now states:

_Kubelet and container runtime version. The [crictl](https://kubernetes.io/docs/tasks/debug/debug-cluster/crictl/) tool can be used to determine if the container runtime supports the feature (crictl info)._

@sftim do you consider this as resolved?

mrunalp commented on keps/sig-node/4033-group-driver-detection-over-cri/README.md:
```diff
@@ -0,0 +1,890 @@
+     // List of current observed runtime conditions.
+     repeated RuntimeCondition conditions = 1;
++    // Configuration settings of the runtime
++    RuntimeConfiguration configuration = 2;
```
I think we should split this into a separate CRI call.

marquiz replied:
@mrunalp I think in this case in this case we should add a streaming interface for the runtime to be able to notify about dynamic changes (not for the cgroup driver but for other potentially more dynamic stuff in the future). Adding yet another endpoint that get's periodically polled doesn't sound reasonable to me. WDYT? @haircommander ?

haircommander replied:
yeah I think something worth noting is we're not preparred in this KEP to react to the runtime changing its mind. I don't think the kubelet should update the cgroup driver if the cri changes without a kubelet restart. from that perspective, separating the calls may be appropriate

marquiz replied:
> I don't think the kubelet should update the cgroup driver if the cri changes

I agree. But I was thinking this from the CRI API pov. So we would add the simple "unary" rpc for now and then later with the dynamic stuff add a parallel streaming rpc (for basically the same data, I think). Is this the way to go?

haircommander replied:
yeah I find myself leaning towards having a separate API call for the streaming stuff, part to reduce the info going over the wire, part to make clear how the kubelet will react to the data

marquiz replied:
OK, changed the proposal to add a new `RuntimeConfig` rpc for querying the runtime configuration. @mrunalp @haircommander WDYT?

marquiz commented on keps/sig-node/4033-group-driver-detection-over-cri/README.md:
```diff
@@ -0,0 +1,890 @@
+     // List of current observed runtime conditions.
+     repeated RuntimeCondition conditions = 1;
++    // Configuration settings of the runtime
++    RuntimeConfiguration configuration = 2;
```
Regarding the runtime configuration we'd prolly need a "system-wide config" that is independent of the runtime class and the runtime handler specific stuff. I think the cgroup driver would fall in the first category – even though at least in containerd it is technically possible to have different cgroup driver for different runtime handlers I can't think of a practical scenario where this would make sense (and kubelet can anyways handle only one)

Also, I think we've tried to prefer lists with named fields over maps also in CRI API.

Thoughts @mrunalp @haircommander @mikebrow ?

AkihiroSuda replied:
>  kubelet can anyways handle only one

👍 

Then this does not need to be a map, but I still expect the proto file to have a comment line to state that this is a global config that is decoupled from individual runtime handlers.

marquiz replied:
I added a comment along these lines in the proposal (in the .proto diff)

aojea commented on keps/sig-node/4033-group-driver-detection-over-cri/README.md:
```diff
@@ -0,0 +1,892 @@
+
+The existing RuntimeStatus message (of the existing Status API endpoint) is
+used as this is being frequently queried by the kubelet, and is a place where
+the runtime tells the Kubelet about its state. The runtime will decide which
```
does it means that the runtime override the flag?

aojea commented on keps/sig-node/4033-group-driver-detection-over-cri/README.md:
```diff
@@ -0,0 +1,892 @@
+
+Kubelet will be modified to support the new field.
+
+If available the cgroup driver information received from the container runtime
+will take precedence over cgroupDriver setting from the kubelet config (or
+`--cgroup-driver` command line flag). If the runtime does not provide
+information about the cgroup driver, then kubelet will fall back to using its
+own configuration (`cgroupDriver` from kubeletConfig or the `--cgroup-driver`
+flag).
+
+Kubelet startup is modified so that connection to the CRI server (container
+runtime) is established and RuntimeStatus is queried before initializing the
+kubelet internal container-manager which is responsible for kubelet-side cgroup
+management.
```
This is an important change of behavior, so if I have kubelet with hardcoded cgroup driver in version X and I upgrade to version X+1 I may have my kubelet running with a different configuration without noticing?

marquiz replied:
I'd argue in this case N+1 would then actually fix your node. Logically there should be only one configuration point for this setting. Having it both in kubelet and container runtime is a constant source of problems as if your runtime has configured driver Y and kubelet has Z then your node is borken

jpbetz commented on keps/sig-node/4033-group-driver-detection-over-cri/README.md:
```diff
@@ -0,0 +1,892 @@
+NOTE: Also set `disable-supported` to `true` or `false` in `kep.yaml`.
+-->
+
+No, though the scope is so small roll back should not be required.
```
nit: This question is asking about feature disablement, not rollback.

marquiz replied:
@jpbetz I now added a feature gate to the proposal so this was changed to 

```
Yes, through the feature gate.
```

Resolved?

jpbetz commented on keps/sig-node/4033-group-driver-detection-over-cri/README.md:
```diff
@@ -0,0 +1,892 @@
+[existing list]: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/
+-->
+
+No feature gate required–the fields are all SIG Node internal and have simple
+fallbacks.
```
Also point out that this feature is only enabled when used with a CRI using the new field? So there is a reasonable level of enablement control.

marquiz replied:
@jpbetz @pacoxu  with the addition of a feature gate this was now changed to 

```
- [X] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: KubeletCgroupDriverFromCRI
  - Components depending on the feature gate: kubelet
```

jpbetz commented on keps/sig-node/4033-group-driver-detection-over-cri/README.md:
```diff
@@ -0,0 +1,892 @@
+automations, so be extremely careful here.
+-->
+
+Yes. If/when the runtime is updated to a version that supports this, kubelet
+will ignore the cgroupDriver config option/flag. However, this change in
+behavior should be largely invisible/irrelevant to the user.
```
This question is primary intended to ensure that features don't make dangerous changes to default behavior.  Speak to that here? Maybe point out that there is no change to default behavior immediately when the `cgroupDriver config option` is not set.

marquiz replied:
This was updated to

```
Yes. If/when the runtime is updated to a version that supports this, kubelet
will ignore the cgroupDriver config option/flag. However, this change in
behavior should not cause any breakages (on the contrary, it should fix
scenarios where the kubelet `--cgorup-driver` setting is incorrectly
configured). With old versions of the container runtimes (that don't support
the new field in the CRI API) the default behavior is not changed.
```

@jpbetz  do you think this is resolved?

mrunalp commented on keps/sig-node/4033-group-driver-detection-over-cri/README.md:
```diff
@@ -0,0 +1,892 @@
+[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
+-->
+
+[ ] I/we understand the owners of the involved components may require updates to
```
This needs to be checked.

mrunalp commented on keps/sig-node/4033-group-driver-detection-over-cri/kep.yaml:
```diff
@@ -0,0 +1,39 @@
+  - sig-node
+status: implementable
+creation-date: 2023-05-25
+reviewers:
```
please fill out reviewers/approvers.

pacoxu commented on keps/sig-node/4033-group-driver-detection-over-cri/README.md:
```diff
@@ -0,0 +1,892 @@
+-->
+
+All CRI implementations support the new cgroupDriver field, and the Kubelet
+drops support for its own CgroupManager field/flag.
```
Probably something like:
- alpha: add cri implementations & add feature gate to auto?
- beta: until e2e is added and FG gets enough feedback
- GA: start the deprecation of cgroup driver flag in kubelet and finally remove it.

marquiz replied:
@pacoxu: this now states:

```
The feature is targeting directly to beta, with the feature gate enabled by
default.

#### Beta

- [ ] Feature implemented, with the feature gate enabled by default.

#### GA

- [ ] released versions of CRI-O and containerd runtime implementations support the feature
- [ ] No bugs reported in the previous cycle.
```

marquiz commented:
Updated:

- changed CRI API to have a new RPC for querying for runtime config (instead of re-using runtime Status)
- update kep.yaml: add reviewers and approvers
- added feature gate (enabled by default)
- changed target matutiry to beta
- add back the deprecation warning about cgroupDriver kubelet config setting
- small updates to PRR

aojea commented on keps/sig-node/4033-group-driver-detection-over-cri/README.md:
```diff
@@ -0,0 +1,914 @@
+If available the cgroup driver information received from the container runtime
+will take precedence over cgroupDriver setting from the kubelet config (or
+`--cgroup-driver` command line flag). If the runtime does not provide
+information about the cgroup driver, then kubelet will fall back to using its
```
this is clear if the response is unimplemented, but what happens if it times out? how long do we wait try?

aojea replied:
I'm fine with anything, just we should document it so we are sure it has been discussed

marquiz replied:
Added the following:

```
If supported by the runtime, RuntimeConfig query is expected to
+succeed, an error (error response or timeout) is regarded as a failed
+initialization of the runtime service and kubelet will exit with an error
+message and an error code.
```

@aojea @haircommander  WDYT?

haircommander replied:
that works with me

haircommander commented on keps/sig-node/4033-group-driver-detection-over-cri/README.md:
```diff
@@ -0,0 +1,914 @@
+ service RuntimeService {
+ ...
++    // RuntimeConfig returns configuration information of the runtime.
++    rpc RuntimeConfig(RuntimeConfigRequest) returns (RuntimeConfigResponse) {}
```
I don't think we need to determine this here, but here's something to bikeshed over: "runtime" is really overloaded. For instance, for @AkihiroSuda 's [RRO](https://github.com/kubernetes/enhancements/pull/3858) proposal, there could be a way for a runtime handler in a runtime implementation to specify runtime features for the oci runtime. 

I would propose we begin to use the phrasing `CRI implementation` and we could begin with `ImplementationConfig`. That opens opportunity for `RuntimeHandlerFeatures` to not conflict too much..

We can always debate this on the implementation PR though, I don't think we need to determine the name concretely before KEP freeze

marquiz replied:
> I don't think we need to determine this here, but here's something to bikeshed over: "runtime" is really overloaded.

Yes, naming is always hard and I concur that it's not essential to try to finalize it here. We can bike-shed over it in the implementation and then update the KEP to reflect that. More important question is the structure of the API.

marquiz commented:
Updated test plan. The proposal is now to only have unit tests for this feature.

johnbelamaric commented on keps/sig-node/4033-group-driver-detection-over-cri/README.md:
```diff
@@ -0,0 +1,928 @@
+- Deprecate the flag
+-->
+
+The feature is targeting directly to beta, with the feature gate enabled by
```
What is the justification for this?

marquiz replied:
Ref: https://github.com/kubernetes/enhancements/pull/4034#discussion_r1215407264

We are confident that this is such a small and small-risk change that it could be enabled by default.

johnbelamaric commented on keps/sig-node/4033-group-driver-detection-over-cri/README.md:
```diff
@@ -0,0 +1,928 @@
+https://github.com/kubernetes/kubernetes/pull/97058/files#diff-7826f7adbc1996a05ab52e3f5f02429e94b68ce6bce0dc534d1be636154fded3R246-R282
+-->
+
+TBD.
```
You can just say they will be written.

marquiz replied:
Changed to

```
Unit tests for the feature gate will be written.
```

johnbelamaric commented on keps/sig-node/4033-group-driver-detection-over-cri/README.md:
```diff
@@ -0,0 +1,928 @@
+that might indicate a serious problem?
+-->
+
+N/A.
```
so, if there were a bug in this code, then a bunch of nodes going NotReady would seem to be a reason to roll back

marquiz replied:
Uh-oh, yes, this was an omission. Updated to:

```
Nodes being in NotReady state with kubelet logs indicating an error in the
RuntimeConfig CRI request, making kubelet fail to start.
```

johnbelamaric commented on keps/sig-node/4033-group-driver-detection-over-cri/README.md:
```diff
@@ -0,0 +1,928 @@
+are missing a bunch of machinery and tooling and can't do that now.
+-->
+
+N/A.
```
Explain why it is not relevant. Maybe something like "there is no persistent state associated with the feature..."

marquiz replied:
Updated to

```
Not planned as there is no persistent state associated with the feature. Manual
testing of the feature gate (in addition to the unit tests) is performed.
```

johnbelamaric commented:
So, this is an important point: https://github.com/kubernetes/enhancements/pull/4034#discussion_r1228475055

I can't approve going straight to beta. I would like to see an alpha phase. If you want an approval today I will need to see that.

I am open to a broader discussion of this with other PRR approvers / TLs / SIG Arch, if you would like, but you will probably need an extension to make that possible.

marquiz commented:
@johnbelamaric thanks for the reviw. We'll think what to do about this.

I'm confident that this is such a simple change that any problems in kubelet (can only show up in kubelet startup) could (and should) be treated and fixed as a "normal" bug.

haircommander commented:
@johnbelamaric I've updated the enhancement to target to alpha. Thanks for your feedback, PTAL

mrunalp commented:
Thanks for the updates.

/approve

johnbelamaric commented on keps/sig-node/4033-group-driver-detection-over-cri/README.md:
```diff
@@ -0,0 +1,942 @@
+logs or events for this purpose.
+-->
+
+Kubelet and container runtime version. The
```
non-blocking comment: since there is a feature gate, the operator can see which nodes it is enabled on via a metric. Please note this in later KEP update.

johnbelamaric reviewed (approved):
/approve
/lgtm
